eth/handler.go

```diff
@@ -824,6 +917,18 @@ func (pm *ProtocolManager) txBroadcastLoop() {
 	}
 }
 
+// hasTx returns an indicator whether local node has the specified
+// transaction no matter in local txpool or chain.
+//
+// Note, this function it's not trivial since read an non-existent
+// data in datbase is expensive.
+func (pm *ProtocolManager) hasTx(hash common.Hash) bool {
+	if pm.txpool.Has(hash) {
+		return true
+	}
+	return rawdb.ReadTxLookupEntry(pm.chaindb, hash) != nil
+}
```

Comment:
Thinking about this, I think we should nuke this function altogether and only ever do txpool.Has.

There are two scenarios that can happen with tx propagation: a) an in-sync node sends an actually new tx and b) a stale node sends us something stale.

In a), which is the usual case, the transaction will either be in the pool already or not yet. Looking into the database is just an extra hit uselessly, since more real transaction propagation is live in-flight transactions.

In b), a stale node could indeed trigger us to download a tx we already have, but that will be caught by the txpool, and a properly implemented node won't keep reannouncing the same old tx over and over again, so apart from some initial wasted bandwidth, we're on track again.

The benefit of dropping this db check clause is that we avoid any potential - deliberate or accidental - DoS vector due to the allowed cheap database access. A malicious node could still feed us bad hashes and force a database lookup after retrieving them, but that also entails them having to incur the bandwidth cost of transferring an entire valid, signed transaction in the first place, so there's a lot less reads they can force us to do.

Reply:
There is also scenario c) where a tx is announced, but by the time we request it it's already mined. In that case we should be able to detect that a block contained the tx and not request it or not be bothered by an empty response. This case should be kind of rare though as it's very improbably for a tx to be announced and immediately included in a block milliseconds later.